### PR TITLE
env/service config param refactor

### PR DESCRIPTION
**Pull request by ryan-cahill**

### CODE REVIEW

File: src/common/docker-compose/index.ts [hunk 1]
```diff
@@ -54,6 +37,31 @@ export const generate = (dependency_manager: DependencyManager): DockerComposeTe
           ...node.parameters,
         },
       };
+
+      for (const [param_name, param_value] of Object.entries(compose.services[node.normalized_ref].environment || {})) {
+        env_params_to_expand[`${node.normalized_ref.toUpperCase()}_${param_name}`.replace(/[.-]/g, '_')] = param_value;
+      }
+      for (const [param_name, param_value] of Object.entries(node.service_config.getParameters())) {
```
Comment (davidthor) [inline src/common/docker-compose/index.ts]: The logic inside this iterator should go into the `getParamValues()` method on the dependency manager. We already parse the service configs there and enrich the key/value parameters on the nodes, so we should be sure to consolidate that logic.

File: src/common/docker-compose/index.ts [hunk 2]
```diff
@@ -87,6 +94,37 @@ export const generate = (dependency_manager: DependencyManager): DockerComposeTe
     }
   });
 
+  const expanded_params = dotenvExpand({ parsed: env_params_to_expand }).parsed;
+  dependency_manager.graph.nodes.forEach(node => {
```
Comment (davidthor) [inline src/common/docker-compose/index.ts]: We're doing something wrong if we're iterating over the nodes twice when we generate the docker compose. Upon complete generation, the graph of nodes and edges should contain all the details needed to do this on a single pass.

File: src/dependency-manager/src/service-config/base.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@ interface RestSubscriptionData {
 
 export interface ServiceParameter {
   description: string;
-  default?: string | number;
+  default?: any;
```
Comment (davidthor) [inline src/dependency-manager/src/service-config/base.ts]: No shortcuts when it comes to the specs we support! Make sure to create a proper interface and add the valueFrom syntax as a valid option.

File: src/common/dependency-manager/local-manager.ts [hunk 4]
```diff
@@ -21,6 +23,92 @@ export default class LocalDependencyManager extends DependencyManager {
     this.config_path = config_path || '';
   }
 
+  protected static loadParameters(dependency_manager: DependencyManager) {
```
Comment (davidthor) [inline src/common/dependency-manager/local-manager.ts]: Any reason this method isn't in the abstract class?

File: src/common/dependency-manager/local-manager.ts [hunk 5]
```diff
@@ -43,6 +131,9 @@ export default class LocalDependencyManager extends DependencyManager {
     // We resolve these after the loop to ensure that explicitly cited service configs take precedence
     await Promise.all(dependency_resolvers.map(fn => fn()));
     dependency_manager.loadSubscriptions();
+    console.log('loading params')
+    this.loadParameters(dependency_manager);
+    console.log('loaded')
```
[the remark this answers is not in the document]
Comment (ryan-cahill) [reply src/common/dependency-manager/local-manager.ts]: :man_facepalming:

File: src/common/dependency-manager/local-manager.ts [hunk 6]
```diff
@@ -21,6 +23,92 @@ export default class LocalDependencyManager extends DependencyManager {
...
+        const service_params = Object.entries(expanded_params || {})
+          .filter(([key, _]) => key.startsWith(service_prefix));
+
+        // reverse order params by length in order to avoid key collisions
+        service_params.sort((pair1: [string, string], pair2: [string, string]) => {
+          return pair2[0].length - pair1[0].length;
+        });
+
+        for (const [param_name, param_value] of service_params) {
+          const real_param_name = param_name.replace(`${service_prefix}_`, '');
```
[the remark this answers is not in the document]
Comment (ryan-cahill) [reply src/common/dependency-manager/local-manager.ts]: That's not entirely true if we want to expand parameters through a tree, indefinitely. In doing this, we also support parameters through the entire tree where a referenced parameter can also be a string that contains a reference to one or many other parameters from the same dependency or datastore

File: src/common/dependency-manager/local-manager.ts (hunk 4 above)
Comment (ryan-cahill) [reply src/common/dependency-manager/local-manager.ts]: The reason was that `LocalServiceNode` isn't accessible to the abstract class. I can move it there, but I didn't think that that really would belong

File: src/common/dependency-manager/local-manager.ts [hunk 7]
```diff
@@ -134,4 +225,32 @@ export default class LocalDependencyManager extends DependencyManager {
     this.graph.addNode(node);
     return node;
   }
+
+  validateNodeDependencies(node: ServiceNode) {
```
[the remark this answers is not in the document]
Comment (ryan-cahill) [reply src/common/dependency-manager/local-manager.ts]: Fair point, I'll add that to the rest of the validation

File: src/commands/push.ts [hunk 8]
```diff
@@ -51,6 +51,7 @@ export default class Push extends Command {
 
     for (const node of dependency_manager.graph.nodes) {
       if (node instanceof LocalServiceNode) {
+        dependency_manager.validateNodeDependencies(node);
```
[the remark this answers is not in the document]
Comment (ryan-cahill) [reply src/commands/push.ts]: I'll update that

[the remark this answers is not in the document]
Comment (ryan-cahill) [reply src/commands/push.ts]: It is now

**Pull request merged**